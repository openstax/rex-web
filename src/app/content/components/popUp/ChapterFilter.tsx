import React from 'react';
import { useIntl } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import AllOrNone from '../../../components/AllOrNone';
import { PlainButton } from '../../../components/Button';
import Checkbox from '../../../components/Checkbox';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import ColorIndicator from '../../highlights/components/ColorIndicator';
import { filters, mobileMarginSides } from '../../styles/PopupConstants';
import { LinkedArchiveTreeNode } from '../../types';
import { splitTitleParts } from '../../utils/archiveTreeUtils';
import { AngleIcon } from './Filters';
import { FiltersChange, LocationFilters } from './types';

// tslint:disable-next-line:variable-name
const Row = styled.div`
  display: flex;
  flex-direction: row;
  ${theme.breakpoints.mobile(css`
    flex-direction: column;
    overflow: hidden;
  `)}
`;

// tslint:disable-next-line:variable-name
const Column = styled.div`
  display: flex;
  flex-direction: column;
  overflow-x: hidden;
`;

// tslint:disable-next-line:variable-name
const ChapterTitle = styled.span`
  display: flex;
  flex-direction: row;
  white-space: nowrap;
  margin-left: 0.8rem;

  > * {
    overflow: hidden;
  }

  .os-text {
    flex: 1;
    text-overflow: ellipsis;
  }

  .os-divider {
    margin: 0 0.4rem;
  }
`;

const chunk = <T extends any>(sections: T[]) => {
  const cutoff = Math.max(20, Math.ceil(sections.length / 2));
  return [sections.slice(0, cutoff), sections.slice(cutoff)].filter((arr) => arr.length > 0);
};

interface ChapterFilterProps {
  ariaLabelItemId?: string;
  className?: string;
  disabled?: boolean;
  locationFilters: LocationFilters;
  locationFiltersWithContent: Map<string, LinkedArchiveTreeNode>;
  selectedLocationFilters: Set<string>;
  multiselect: boolean;
  setFilters: (filters: FiltersChange<LinkedArchiveTreeNode>) => void;
}

// tslint:disable-next-line:variable-name
const ChapterFilter = (props: ChapterFilterProps) => {
  const [openChapterId, setOpenChapterId] = React.useState<string | null>(null);
  const intl = useIntl();

  React.useEffect(() => {
    const selectedSectionId = Array.from(props.selectedLocationFilters).pop();
    if (selectedSectionId) {
      const filterWithSelectedSection = Array.from(props.locationFilters.values()).find(({ children }) => {
        return children && children.find((section) => section.id === selectedSectionId);
      });
      if (filterWithSelectedSection) {
        setOpenChapterId(filterWithSelectedSection.section.id);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const setSelectedChapters = (change: FiltersChange<LinkedArchiveTreeNode>) => {
    props.setFilters(change);
  };

  const handleChange = (section: LinkedArchiveTreeNode) => {
    if (props.selectedLocationFilters.has(section.id)) {
      setSelectedChapters({ remove: [section], new: [] });
    } else {
      setSelectedChapters({ remove: [], new: [section] });
    }
  };

  const getAriaLabel = (section: LinkedArchiveTreeNode) => {
    if (props.ariaLabelItemId) {
      return intl.formatMessage({ id: props.ariaLabelItemId }, { filter: splitTitleParts(section.title).join(' ') });
    }
  };

  const values = Array.from(props.locationFilters.values());
  const hasFiltersWithChildren = Boolean(values.find((filter) => filter.children));
  const sectionChunks = hasFiltersWithChildren ? [values] : chunk(values);

  return <div className={props.className} tabIndex={-1}>
    {props.multiselect
      ? (
        <AllOrNone
          onNone={() => setSelectedChapters({ remove: Array.from(props.locationFiltersWithContent.values()), new: [] })}
          onAll={() => setSelectedChapters({ remove: [], new: Array.from(props.locationFiltersWithContent.values()) })}
          disabled={props.disabled}
        />
      )
      : null}
    <Row>
      {sectionChunks.map((sectionChunk, index) => <Column key={index}>
        {sectionChunk.map((location) => {
          const { section, children } = location;
          if (!children) {
            return <ChapterFilterItem
              key={section.id}
              selected={props.selectedLocationFilters.has(section.id)}
              disabled={props.disabled || !props.locationFiltersWithContent.has(section.id)}
              multiselect={Boolean(props.multiselect)}
              title={section.title}
              onChange={() => handleChange(section)}
              ariaLabel={getAriaLabel(section)}
              dataAnalyticsLabel={`Filter PQ by ${splitTitleParts(section.title).join(' ')}`}
            />;
          } else {
            return <StyledDetails key={section.id} open={openChapterId === section.id}>
              <StyledSummary onClick={(ev: React.MouseEvent) => {
                ev.preventDefault();
                setOpenChapterId((currentId) => currentId !== section.id ? section.id : null);
              }}>
                <ChapterTitle dangerouslySetInnerHTML={{__html: section.title}} />
                <AngleIcon direction={openChapterId === section.id ? 'up' : 'down'} />
              </StyledSummary>
              <StyledChapterFilterItemWrapper>
                {children.map((child) => (
                  <ChapterFilterItem
                    key={child.id}
                    selected={props.selectedLocationFilters.has(child.id)}
                    disabled={false}
                    multiselect={props.multiselect}
                    title={child.title}
                    onChange={() => handleChange(child)}
                    ariaLabel={getAriaLabel(child)}
                    dataAnalyticsLabel={`Filter PQ by ${splitTitleParts(child.title).join(' ')}`}
                  />
                ))}
              </StyledChapterFilterItemWrapper>
            </StyledDetails>;
          }
        })}
      </Column>)}
    </Row>
  </div>;
};

interface ChapterFilterItemProps {
  selected: boolean;
  disabled: boolean;
  multiselect: boolean;
  title: string;
  ariaLabel?: string;
  dataAnalyticsLabel: string;
  onChange: () => void;
}

// tslint:disable-next-line: variable-name
const ChapterFilterItem = (props: ChapterFilterItemProps) => {
  if (props.multiselect) {
    return <Checkbox
      checked={props.selected}
      disabled={props.disabled}
      onChange={props.onChange}
      aria-label={props.ariaLabel}
    >
      <ChapterTitle dangerouslySetInnerHTML={{__html: props.title}} />
    </Checkbox>;
  }

  return <StyledSectionItem
    onClick={props.onChange}
    isSelected={props.selected}
    aria-label={props.ariaLabel}
    data-analytics-label={props.dataAnalyticsLabel}
  >
    <ChapterTitle dangerouslySetInnerHTML={{__html: props.title}} />
  </StyledSectionItem>;
};

// tslint:disable-next-line: variable-name
export const StyledDetails = styled.details`
  width: 400px;
  cursor: pointer;
  border-bottom: 1px solid ${theme.color.neutral.formBorder};
  overflow: visible;
  ${theme.breakpoints.mobileSmall(css`
    width: 100%;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StyledSummary = styled.summary`
  padding: 1rem 1.6rem;
  list-style: none;

  &::marker,
  &::-webkit-details-marker {
    content: "";
    display: none;
  }

  ${ChapterTitle} {
    float: left;
    margin-left: 0;
    text-align: left;
  }

  ${AngleIcon} {
    float: right;
  }
`;

// tslint:disable-next-line: variable-name
export const StyledSectionItem = styled(PlainButton)`
  display: flex;
  align-items: center;
  height: 4rem;
  width: 100%;
  text-align: left;
  font-size: 1.4rem;
  ${textStyle}
  ${(props: { isSelected: boolean }) => {
    if (props.isSelected) {
      return 'color: #027eb5;';
    }
  }}

  &:hover,
  &:focus {
    background-color: ${theme.color.neutral.pageBackground};
  }

  ${ChapterTitle} {
    padding-left: 2.5rem;
    width: 95%;
  }
`;

// tslint:disable-next-line: variable-name
export const StyledChapterFilterItemWrapper = styled.div`
  overflow: visible;
`;

export default styled(ChapterFilter)`
  ${textStyle}
  background: ${theme.color.white};
  font-size: 1.4rem;
  padding: ${filters.dropdownContent.padding.topBottom}rem ${filters.dropdownContent.padding.sides}rem;
  outline: none;
  overflow: auto;
  z-index: 1;

  ${AllOrNone} {
    margin: 0.8rem 0 0.8rem 0.8rem;
  }

  ${Checkbox} {
    padding: 0.8rem;
  }

  ${ColorIndicator} {
    margin: 0 1.6rem 0 1.6rem;
  }

  ${StyledDetails} {
    &:last-child {
      border-bottom: none;
    }
  }

  ${theme.breakpoints.mobileSmall(css`
    width: calc(100vw - ${mobileMarginSides * 2}rem);
  `)}
`;
