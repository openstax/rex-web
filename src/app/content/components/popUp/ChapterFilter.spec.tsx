import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import { book as archiveBook, page } from '../../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../../test/mocks/osWebLoader';
import TestContainer from '../../../../test/TestContainer';
import { receiveLoggedOut } from '../../../auth/actions';
import AllOrNone from '../../../components/AllOrNone';
import { ButtonLink } from '../../../components/Button';
import Checkbox from '../../../components/Checkbox';
import { Store } from '../../../types';
import { assertDefined } from '../../../utils';
import { receiveBook, receivePage } from '../../actions';
import { receiveHighlightsTotalCounts } from '../../highlights/actions';
import { ConnectedChapterFilter } from '../../highlights/components/SummaryPopup/Filters';
import { receiveStudyGuidesTotalCounts } from '../../studyGuides/actions';
import Filters from '../../studyGuides/components/Filters';
import { formatBookData, stripIdVersion } from '../../utils';
import { findArchiveTreeNodeById } from '../../utils/archiveTreeUtils';
import ChapterFilter, { StyledDetails, StyledSummary } from './ChapterFilter';
import { LocationFilters } from './types';

describe('ChapterFilter', () => {
  const book = formatBookData(archiveBook, mockCmsBook);
  const locationIds = new Map() as LocationFilters;
  let store: Store;

  beforeEach(() => {
    store = createTestStore();

    store.dispatch(receivePage({...page, references: []}));
  });

  it('matches snapshot', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testpage1-uuid': {[HighlightColorEnum.Green]: 1},
    }, new Map([[
      'testbook1-testpage1-uuid',
      { section: assertDefined(findArchiveTreeNodeById(book.tree, 'testbook1-testpage1-uuid'), '') },
    ]])));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without a book', () => {
    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const checkedBoxes = component.root.findAllByProps({checked: true});
    expect(checkedBoxes.length).toBe(0);
  });

  it('initially has selected chapters with highlights', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testchapter3-uuid': {[HighlightColorEnum.Green]: 3},
      'testbook1-testpage1-uuid': {[HighlightColorEnum.Pink]: 1},
    }, new Map([
      [
        'testbook1-testpage1-uuid',
        { section: assertDefined(findArchiveTreeNodeById(book.tree, 'testbook1-testpage1-uuid'), '') },
      ],
      [
        'testbook1-testchapter3-uuid',
        { section: assertDefined(findArchiveTreeNodeById(book.tree, 'testbook1-testchapter3-uuid'), '') },
      ],
    ])));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const [box1, box2, box3, box4, box5] = component.root.findAllByType(Checkbox);

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(false);
    expect(box3.props.checked).toBe(false);
    expect(box4.props.checked).toBe(false);
    expect(box5.props.checked).toBe(true);
  });

  it('checks and unchecks chapters', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testpage1-uuid': {[HighlightColorEnum.Green]: 1},
    }, new Map([[
      'testbook1-testpage1-uuid',
      { section: assertDefined(findArchiveTreeNodeById(book.tree, 'testbook1-testpage1-uuid'), '') },
    ]])));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const [box1] = component.root.findAllByType(Checkbox);

    expect(box1.props.checked).toBe(true);

    renderer.act(() => {
      box1.props.onChange();
    });

    expect(box1.props.checked).toBe(false);

    renderer.act(() => {
      box1.props.onChange();
    });

    expect(box1.props.checked).toBe(true);
  });

  it('selects none', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testpage1-uuid': {[HighlightColorEnum.Green]: 1},
    }, new Map([[
      'testbook1-testpage1-uuid',
      { section: assertDefined(findArchiveTreeNodeById(book.tree, 'testbook1-testpage1-uuid'), '') },
    ]])));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const [box1, box2] = component.root.findAllByType(Checkbox);
    const allOrNone = component.root.findByType(AllOrNone);

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(false);

    renderer.act(() => {
      allOrNone.props.onNone();
    });

    expect(box1.props.checked).toBe(false);
    expect(box2.props.checked).toBe(false);
  });

  it('selects all select only chapters with highlights', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testchapter3-uuid': {[HighlightColorEnum.Green]: 3},
      'testbook1-testpage1-uuid': {[HighlightColorEnum.Green]: 1},
    }, locationIds));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const [box1, box2, , , box5] = component.root.findAllByType(Checkbox);
    const allOrNone = component.root.findByType(AllOrNone);

    renderer.act(() => {
      allOrNone.props.onNone();
    });

    expect(box1.props.checked).toBe(false);
    expect(box2.props.checked).toBe(false);
    expect(box5.props.checked).toBe(false);

    renderer.act(() => {
      allOrNone.props.onAll();
    });

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(false);
    expect(box5.props.checked).toBe(true);
  });

  it('chapters without highlights are disabled', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testpage1-uuid': {[HighlightColorEnum.Green]: 1},
    }, locationIds));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedChapterFilter multiselect={true} />
    </TestContainer>);

    const [box1, ...otherBoxes] = component.root.findAllByType(Checkbox);

    expect(box1.props.disabled).toBe(false);
    expect(otherBoxes.every((box) => box.props.disabled)).toBe(true);
  });

  it('disables all or none when logged out user', () => {
    store.dispatch(receiveLoggedOut());
    const pageId = stripIdVersion(book.tree.contents[0].id);
    store.dispatch(receiveStudyGuidesTotalCounts({
      [pageId]: {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Yellow]: 1,
      },
    }));

    const component = renderer.create(<TestContainer store={store}>
        <Filters />
      </TestContainer>);

    const [...allOrNoneButtons] = component.root.findAllByType(ButtonLink);
    expect(allOrNoneButtons.every((button) => button.props.disabled)).toBe(true);
  });

  it('toggle <details> on click', () => {
    const locationFilters = new Map([[
      'testbook1-testchapter2-uuid',
      {
        children: [{ id: 'testbook1-testpage3-uuid', title: 'page' }],
        section: { id: 'testbook1-testchapter2-uuid', title: 'chapter' },
      },
    ]]);

    const component = renderer.create(<TestContainer store={store}>
      <ChapterFilter
        locationFilters={locationFilters}
        locationFiltersWithContent={new Set()}
        selectedLocationFilters={new Set()}
        ariaLabelItemId='i18n:practice-questions:popup:filters:filter-by:aria-label'
      />
    </TestContainer>);

    const [details] = component.root.findAllByType(StyledDetails);
    const [summary] = details.findAllByType(StyledSummary);
    expect(details.props.open).toEqual(false);

    renderer.act(() => {
      summary.props.onClick({ preventDefault: jest.fn() });
    });

    component.update(<TestContainer store={store}>
      <ChapterFilter
        locationFilters={locationFilters}
        locationFiltersWithContent={new Set()}
        selectedLocationFilters={new Set()}
        ariaLabelItemId='i18n:practice-questions:popup:filters:filter-by:aria-label'
      />
    </TestContainer>);

    expect(details.props.open).toEqual(true);
  });
});
