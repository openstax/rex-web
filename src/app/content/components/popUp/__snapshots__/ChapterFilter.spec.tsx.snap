// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChapterFilter matches snapshot 1`] = `
<div
  className="ChapterFilter-tcdlu-7 iaQiZB"
  tabIndex={-1}
>
  <div
    className="AllOrNone-sc-7m1y94-0 divvKQ"
  >
    <button
      className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 flLYdq"
      onClick={[Function]}
    >
      All
    </button>
    <span>
      |
    </span>
    <button
      className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 flLYdq"
      onClick={[Function]}
    >
      None
    </button>
  </div>
  <div
    className="ChapterFilter__Row-tcdlu-0 nqIYu"
  >
    <div
      className="ChapterFilter__Column-tcdlu-1 dKPjiE"
    >
      <label
        className="Checkbox-f0dje9-2 ksczzL"
      >
        <input
          checked={true}
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-text\\">Test Page 1</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1.1</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">2</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 2</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">3</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 3</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">4</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 4</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">10</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 5</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">12</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 6</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">A</span><span class=\\"os-divider\\"> | </span><span class=\\"os-text\\">Atomic Masses</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-number\\">D</span><span class=\\"os-divider\\"> | </span><span class=\\"os-text\\">Glossary of Key Symbols and Notation</span>",
            }
          }
        />
      </label>
      <label
        className="Checkbox-f0dje9-2 gEkNQh"
      >
        <input
          checked={false}
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
        >
          <svg
            aria-hidden="true"
            className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          className="ChapterFilter__ChapterTitle-tcdlu-2 fmaoPu"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<span class=\\"os-text\\">Test Page for Generic Styles</span>",
            }
          }
        />
      </label>
    </div>
  </div>
</div>
`;
